### ISSUE TITLE
`regex_creation_in_loops` fires even if a regex compilation is followed by `break`.

### ISSUE THREAD

**Issue by your-diary** [rust-lang/rust-clippy#15484]
### Summary

`regex_creation_in_loops` is blindly fired even if a regex compilation is followed by `break`, whilst `break` makes the number of times the compilation occurs exactly one.

Here's a real use-case in which I think the lint should not be fired (I also provide a minimal working example in the next section):

```rust
let stdout = BufReader::new(child.stdout.take().unwrap());

let mut authorization_code = None;

for line in stdout.lines() { // iterates the lines of stdout of a child process
    let line = line?;
    if line.contains(callback_endpoint) { // if a line satisfies some condition
        #[allow(clippy::regex_creation_in_loops)]
        let r = Regex::new(r#"\?code=([^&]+)&"#)?; // we compile a regex once
        authorization_code = Some( // does some work using the regex
            r.captures(&line)
                .unwrap()
                .get(1)
                .unwrap()
                .as_str()
                .to_owned(),
        );
        break; // and break from the loop
    }
}
```

This forces me to add `#[allow]` attribute or unnecessarily broaden the scope of a regex. (In particular the latter is unacceptable.)

### Lint Name

regex_creation_in_loops

### Reproducer

I tried this code ([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=02c286c7cc051f3fcef80a97de30cac8)):

```rust
use regex::Regex;

fn main() {
    let some_condition = true;

    loop {
        if some_condition {
            let _ = Regex::new(r#".*"#).unwrap();
            break;
        }
    }
}
```

I saw this happen:

```
warning: compiling a regex in a loop
 --> src/main.rs:8:21
  |
8 |             let _ = Regex::new(r#".*"#).unwrap();
  |                     ^^^^^^^^^^
  |
help: move the regex construction outside this loop
 --> src/main.rs:6:5
  |
6 |     loop {
  |     ^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#regex_creation_in_loops
  = note: `#[warn(clippy::regex_creation_in_loops)]` on by default
```

I expected to see this happen:

- No warning should be shown.

- Generally, branch analysis should be performed to remove this kind of false positives.

    ```rust
    // the lint should be fired
    loop {
        let _ = Regex::new(r#".*"#).unwrap();
    }

    // the lint should NOT be fired
    loop {
        if some_condition {
            let _ = Regex::new(r#".*"#).unwrap();
            break;
        }
    }

    // the lint should be fired
    loop {
        if some_condition {
            let _ = Regex::new(r#".*"#).unwrap();
            if another_condition {
                continue;
            }
            break;
        }
    }
    ```

### Version

```text
rustc 1.89.0 (29483883e 2025-08-04)
binary: rustc
commit-hash: 29483883eed69d5fb4db01964cdf2af4d86e9cb2
commit-date: 2025-08-04
host: aarch64-apple-darwin
release: 1.89.0
LLVM version: 20.1.7
```

**Comment by ada4a** [comment]
The example looks a bit fragile either way if I'm being completely honest.. one hasty refactor and you're back to creating the regex in the loop. Wouldn't a [`LazyCell`](https://doc.rust-lang.org/std/cell/struct.LazyCell.html) (created outside the loop) be safer here anyway?
